### ISSUE TITLE
Freeing Memory/Compaction Issue

### ISSUE THREAD

**Issue by davidpean** [influxdata/influxdb#13209]
**Version**: 1.7.4
**OS**: Amazon Linux
**Series Count**: 7.3M
**Writes Per Second**: 50-60k

I have two Influx databases taking the same data, and one of them seemed to have gotten in a bad state a few days ago. Normally the memory drops and rises throughout the day. 

<img width="1374" alt="Screen Shot 2019-04-05 at 4 15 51 PM" src="https://user-images.githubusercontent.com/317856/55654144-93684e00-57be-11e9-9dce-3dd17c4ca85c.png">

The last several days it has not dropped, and queries running on it will spike up the CPU considerably thus affecting write speed. When queries are removed, the writes are OK but the memory doesn't seem to budge. The only other correlation I can make is that the PauseTotalNS time also picked up during the same time

<img width="1366" alt="Screen Shot 2019-04-05 at 4 21 27 PM" src="https://user-images.githubusercontent.com/317856/55654258-e3dfab80-57be-11e9-8999-73193420f0d0.png">

Garbage collection time 

go_gc_duration_seconds{quantile="1"} 3.491141662

Full Metrics:

```
# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 0.001018825
go_gc_duration_seconds{quantile="0.25"} 0.004062893
go_gc_duration_seconds{quantile="0.5"} 0.012319033
go_gc_duration_seconds{quantile="0.75"} 0.03302532
go_gc_duration_seconds{quantile="1"} 3.491141662
go_gc_duration_seconds_sum 3853.941777687
go_gc_duration_seconds_count 376118
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 147
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.11"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 2.0927280664e+10
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 5.03310507218012e+15
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 4.7580958e+07
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 3.4314028099787e+13
# HELP go_memstats_gc_cpu_fraction The fraction of this program's available CPU time used by the GC since the program started.
# TYPE go_memstats_gc_cpu_fraction gauge
go_memstats_gc_cpu_fraction 0.02427343321728035
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 4.67017728e+09
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 2.0927280664e+10
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 9.3189545984e+10
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 3.04805888e+10
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 1.26228595e+08
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 7.6815630336e+10
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 1.23670134784e+11
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.5544944845373735e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 3.4314154328382e+13
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 110592
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 131072
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 5.80618568e+08
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 1.341177856e+09
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 2.8892934096e+10
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.45315346e+08
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 1.0747904e+07
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 1.0747904e+07
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 1.298852652e+11
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 141
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 6.785625458e+07
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 100000
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 3369
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 4.06371332096e+11
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.55007886782e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 2.202656538624e+12
```

**Comment by davidpean** [comment]
Not sure if this is related to above, but another problem has arisen. There seems to have been a problem with the last few shards, where compaction is either stuck or taking too long. The size of the data on disk has skyrocketed the last few days (6TB total disk)

<img width="1043" alt="Screen Shot 2019-04-14 at 5 39 23 PM" src="https://user-images.githubusercontent.com/317856/56100230-058a1280-5ee4-11e9-94c5-2e01cec7da6d.png">

I came across this https://github.com/influxdata/influxdb/issues/10465# (which has been closed) but found some helpful commands and internal metrics to see what was going on with the shards. It looks like yesterdays shard is still doing a full compaction, while the current is also having problems.

`find /influx/data/kafka_metrics_db/ -name "*.tsm.tmp"
/influx/data/kafka_metrics_db/two_weeks/123/000011632-000000005.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011712-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000007784-000000004.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000007752-000000004.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011752-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011768-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000007720-000000004.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011776-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011744-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011736-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011784-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000007688-000000004.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011760-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000007816-000000004.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000007560-000000006.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011720-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/123/000011728-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003456-000000012.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009416-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009456-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000010248-000000006.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003456-000000007.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000011456-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003456-000000006.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000010248-000000007.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003784-000000003.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003456-000000008.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009432-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009424-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009408-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003752-000000003.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003688-000000006.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003456-000000010.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009440-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009464-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003456-000000009.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003456-000000011.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000003688-000000005.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000011520-000000001.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009472-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000011521-000000001.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000009448-000000002.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000011522-000000001.tsm.tmp
/influx/data/kafka_metrics_db/two_weeks/125/000011523-000000001.tsm.tmp`

Yesterdays shard (id=123) has compactions still going on

`> select /.*CompactionsActive/ from tsm1_engine where time > now() - 1m AND id = '123' group by *
name: tsm1_engine
tags: database=kafka_metrics_db, engine=tsm1, hostname=ip-10-141-29-141, id=123, indexType=tsi1, path=/influx/data/kafka_metrics_db/two_weeks/123, retentionPolicy=two_weeks, walPath=/db/kafka_metrics_db/two_weeks/123
time                cacheCompactionsActive tsmFullCompactionsActive tsmLevel1CompactionsActive tsmLevel2CompactionsActive tsmLevel3CompactionsActive tsmOptimizeCompactionsActive
----                ---------------------- ------------------------ -------------------------- -------------------------- -------------------------- ----------------------------
1555277830000000000 0                      1                        10                         5                          0                          0
1555277840000000000 0                      1                        10                         5                          0                          0
1555277850000000000 0                      1                        10                         5                          0                          0
1555277860000000000 0                      1                        11                         5                          0                          0
1555277870000000000 0                      1                        11                         5                          0                          0
1555277880000000000 0                      1                        11                         5                          0                          0`

Digging into the shard sizes, it appears to be dropping slowly, but are way larger than usual. The last 7 days of shard size:

<img width="1053" alt="Screen Shot 2019-04-14 at 6 32 40 PM" src="https://user-images.githubusercontent.com/317856/56100262-82b58780-5ee4-11e9-982b-3db888313b5b.png">

**Comment by davidpean** [comment]
output of http://localhost:8086/debug/pprof/profile

[influx-profile.pprof.zip](https://github.com/influxdata/influxdb/files/3077947/influx-profile.pprof.zip)

**Comment by davidpean** [comment]
"shard:/influx/data/kafka_metrics_db/two_weeks/123:123": {"name":"shard","tags":{"database":"kafka_metrics_db","engine":"tsm1","id":"123","indexType":"tsi1","path":"/influx/data/kafka_metrics_db/two_weeks/123","retentionPolicy":"two_weeks","walPath":"/db/kafka_metrics_db/two_weeks/123"},"values":{"diskBytes":1888591274500,"fieldsCreate":46015,"seriesCreate":2024872,"writeBytes":0,"writePointsDropped":0,"writePointsErr":0,"writePointsOk":4753483531,"writeReq":969303,"writeReqErr":0,"writeReqOk":969303}},
"tsm1_engine:/influx/data/kafka_metrics_db/two_weeks/123:123": {"name":"tsm1_engine","tags":{"database":"kafka_metrics_db","engine":"tsm1","id":"123","indexType":"tsi1","path":"/influx/data/kafka_metrics_db/two_weeks/123","retentionPolicy":"two_weeks","walPath":"/db/kafka_metrics_db/two_weeks/123"},"values":{"cacheCompactionDuration":86472821208003,"cacheCompactionErr":0,"cacheCompactions":3586,"cacheCompactionsActive":0,"tsmFullCompactionDuration":137658955388281,"tsmFullCompactionErr":0,"tsmFullCompactionQueue":0,"tsmFullCompactions":28,"tsmFullCompactionsActive":2,"tsmLevel1CompactionDuration":2014561570338649,"tsmLevel1CompactionErr":0,"tsmLevel1CompactionQueue":203,"tsmLevel1Compactions":1505,"tsmLevel1CompactionsActive":9,"tsmLevel2CompactionDuration":445626378136675,"tsmLevel2CompactionErr":0,"tsmLevel2CompactionQueue":100,"tsmLevel2Compactions":268,"tsmLevel2CompactionsActive":7,"tsmLevel3CompactionDuration":63049740211776,"tsmLevel3CompactionErr":0,"tsmLevel3CompactionQueue":0,"tsmLevel3Compactions":28,"tsmLevel3CompactionsActive":0,"tsmOptimizeCompactionDuration":0,"tsmOptimizeCompactionErr":0,"tsmOptimizeCompactionQueue":0,"tsmOptimizeCompactions":0,"tsmOptimizeCompactionsActive":0}},
"tsm1_cache:/influx/data/kafka_metrics_db/two_weeks/123:123": {"name":"tsm1_cache","tags":{"database":"kafka_metrics_db","engine":"tsm1","id":"123","indexType":"tsi1","path":"/influx/data/kafka_metrics_db/two_weeks/123","retentionPolicy":"two_weeks","walPath":"/db/kafka_metrics_db/two_weeks/123"},"values":{"WALCompactionTimeMs":86470472,"cacheAgeMs":81874999,"cachedBytes":5855722379373,"diskBytes":0,"memBytes":0,"snapshotCount":0,"writeDropped":0,"writeErr":0,"writeOk":969303}},
"tsm1_filestore:/influx/data/kafka_metrics_db/two_weeks/123:123": {"name":"tsm1_filestore","tags":{"database":"kafka_metrics_db","engine":"tsm1","id":"123","indexType":"tsi1","path":"/influx/data/kafka_metrics_db/two_weeks/123","retentionPolicy":"two_weeks","walPath":"/db/kafka_metrics_db/two_weeks/123"},"values":{"diskBytes":1888591274500,"numFiles":2568}},
"tsm1_wal:/influx/data/kafka_metrics_db/two_weeks/123:123": {"name":"tsm1_wal","tags":{"database":"kafka_metrics_db","engine":"tsm1","id":"123","indexType":"tsi1","path":"/influx/data/kafka_metrics_db/two_weeks/123","retentionPolicy":"two_weeks","walPath":"/db/kafka_metrics_db/two_weeks/123"},"values":{"currentSegmentDiskBytes":0,"oldSegmentsDiskBytes":0,"writeErr":0,"writeOk":969303}},

**Comment by davidpean** [comment]
I tried to do a restart. It took about 3 hours+ since it had to read from two very large shards. The same behavior of the new shard that started within this time frame had similar behavior. I had to delete all the files in the data directories and start over. I will see how it performs over the next day

**Comment by fenneh** [comment]
hey @davidpean - Have you experienced this since?

I'm currently seeing very similar behaviour where shards are appearing to be not compacting and growing to massive sizes e.g. 300/400gb instead of 100gb and it happens back to back. 

We've combatted it previously by shipping the data off the box, loading it up into another influx, exporting and then re-importing into the original host which is a very very time consuming process. We've also dropped our shard durations down to minimize the data loss we seem to having now due to it re-occuring.
